File: go/internal/mldsa/hash.go
package mldsa

import (
	"encoding/binary"

	"golang.org/x/crypto/sha3"
)

// H computes SHAKE-256 hash with the given output length.
func H(data []byte, outLen int) []byte {
	h := sha3.NewShake256()
	h.Write(data)
	out := make([]byte, outLen)
	h.Read(out)
	return out
}

// ExpandA generates the k x l matrix A from seed rho using SHAKE-128.
// FIPS 204 Algorithm 32 (ExpandA) / Algorithm 30 (RejNTTPoly).
func ExpandA(rho []byte, k, l int) [][]([256]int) {
	A := make([][]([256]int), k)
	for i := 0; i < k; i++ {
		A[i] = make([][256]int, l)
		for j := 0; j < l; j++ {
			A[i][j] = rejNTTPoly(rho, byte(j), byte(i))
		}
	}
	return A
}

// rejNTTPoly generates a single polynomial via rejection sampling from SHAKE-128.
// FIPS 204 Algorithm 30.
func rejNTTPoly(rho []byte, j, i byte) [256]int {
	// Domain separator: rho || IntegerToBytes(s, 2) where s = 256*i + j
	// Per FIPS 204, the two bytes are j then i (little-endian of 256*i+j).
	seed := make([]byte, len(rho)+2)
	copy(seed, rho)
	seed[len(rho)] = j
	seed[len(rho)+1] = i

	h := sha3.NewShake128()
	h.Write(seed)

	var poly [256]int
	idx := 0
	buf := make([]byte, 3)
	for idx < 256 {
		h.Read(buf)
		// CoeffFromThreeBytes
		b0 := int(buf[0])
		b1 := int(buf[1])
		b2 := int(buf[2])
		val := b0 | (b1 << 8) | (b2 << 16)
		val &= 0x7FFFFF // mask to 23 bits
		if val < Q {
			poly[idx] = val
			idx++
		}
	}
	return poly
}

// ExpandS generates vectors s1 (length l) and s2 (length k) with coefficients
// in [-eta, eta] from seed sigma.
// FIPS 204 Algorithm 33 (ExpandS) / Algorithm 31 (RejBoundedPoly).
func ExpandS(sigma []byte, eta, k, l int) ([][256]int, [][256]int) {
	s1 := make([][256]int, l)
	s2 := make([][256]int, k)
	for r := 0; r < l; r++ {
		s1[r] = rejBoundedPoly(sigma, eta, uint16(r))
	}
	for r := 0; r < k; r++ {
		s2[r] = rejBoundedPoly(sigma, eta, uint16(l+r))
	}
	return s1, s2
}

// rejBoundedPoly samples a polynomial with coefficients in [-eta, eta].
// FIPS 204 Algorithm 31.
func rejBoundedPoly(sigma []byte, eta int, nonce uint16) [256]int {
	seed := make([]byte, len(sigma)+2)
	copy(seed, sigma)
	binary.LittleEndian.PutUint16(seed[len(sigma):], nonce)

	h := sha3.NewShake256()
	h.Write(seed)

	var poly [256]int
	idx := 0
	buf := make([]byte, 1)
	for idx < 256 {
		h.Read(buf)
		z := int(buf[0])
		z0 := z & 0x0F
		z1 := z >> 4

		if eta == 2 {
			if z0 < 15 {
				val := z0 - (z0*205>>10)*5 // z0 mod 5
				poly[idx] = ModQ(int64(eta - val))
				idx++
			}
			if idx < 256 && z1 < 15 {
				val := z1 - (z1*205>>10)*5 // z1 mod 5
				poly[idx] = ModQ(int64(eta - val))
				idx++
			}
		} else if eta == 4 {
			if z0 < 9 {
				poly[idx] = ModQ(int64(eta - z0))
				idx++
			}
			if idx < 256 && z1 < 9 {
				poly[idx] = ModQ(int64(eta - z1))
				idx++
			}
		}
	}
	return poly
}

// ExpandMask generates a vector y of l polynomials with coefficients in
// [-(gamma1-1), gamma1].
// FIPS 204 Algorithm 34 (ExpandMask).
func ExpandMask(rhoPrime []byte, gamma1, l, kappa int) [][256]int {
	y := make([][256]int, l)
	for r := 0; r < l; r++ {
		nonce := uint16(kappa + r)
		seed := make([]byte, len(rhoPrime)+2)
		copy(seed, rhoPrime)
		binary.LittleEndian.PutUint16(seed[len(rhoPrime):], nonce)

		h := sha3.NewShake256()
		h.Write(seed)

		var bitLen int
		if gamma1 == (1 << 17) {
			bitLen = 18
		} else {
			bitLen = 20
		}
		byteCount := 256 * bitLen / 8
		buf := make([]byte, byteCount)
		h.Read(buf)

		y[r] = bitUnpackGamma1(buf, gamma1, bitLen)
	}
	return y
}

// bitUnpackGamma1 unpacks bytes into polynomial coefficients in [-(gamma1-1), gamma1].
func bitUnpackGamma1(buf []byte, gamma1, bitLen int) [256]int {
	var poly [256]int
	if bitLen == 18 {
		// 18 bits per coefficient, 4 coefficients per 9 bytes
		for i := 0; i < 64; i++ {
			b := buf[9*i : 9*i+9]
			vals := [4]int{
				int(b[0]) | int(b[1])<<8 | (int(b[2])&0x03)<<16,
				int(b[2])>>2 | int(b[3])<<6 | (int(b[4])&0x0F)<<14,
				int(b[4])>>4 | int(b[5])<<4 | (int(b[6])&0x3F)<<12,
				int(b[6])>>6 | int(b[7])<<2 | int(b[8])<<10,
			}
			for j := 0; j < 4; j++ {
				vals[j] &= (1 << 18) - 1
				poly[4*i+j] = ModQ(int64(gamma1 - vals[j]))
			}
		}
	} else {
		// 20 bits per coefficient, 4 coefficients per 10 bytes
		for i := 0; i < 64; i++ {
			b := buf[10*i : 10*i+10]
			vals := [4]int{
				int(b[0]) | int(b[1])<<8 | (int(b[2])&0x0F)<<16,
				int(b[2])>>4 | int(b[3])<<4 | (int(b[4])&0xFF)<<12,
				int(b[5]) | int(b[6])<<8 | (int(b[7])&0x0F)<<16,
				int(b[7])>>4 | int(b[8])<<4 | int(b[9])<<12,
			}
			for j := 0; j < 4; j++ {
				vals[j] &= (1 << 20) - 1
				poly[4*i+j] = ModQ(int64(gamma1 - vals[j]))
			}
		}
	}
	return poly
}

// SampleInBall generates the challenge polynomial c with exactly tau nonzero
// coefficients, each +/- 1, using Fisher-Yates shuffle.
// FIPS 204 Algorithm 29.
func SampleInBall(cTilde []byte, tau int) [256]int {
	var c [256]int

	h := sha3.NewShake256()
	h.Write(cTilde)

	// Read 8 bytes for sign bits
	signBytes := make([]byte, 8)
	h.Read(signBytes)
	signs := binary.LittleEndian.Uint64(signBytes)

	buf := make([]byte, 1)
	for i := 256 - tau; i < 256; i++ {
		// Sample j uniformly from [0, i]
		for {
			h.Read(buf)
			j := int(buf[0])
			if j <= i {
				c[i] = c[j]
				if signs&1 == 0 {
					c[j] = 1
				} else {
					c[j] = ModQ(int64(-1)) // Q - 1
				}
				signs >>= 1
				break
			}
		}
	}
	return c
}

File: js/src/mldsa/hash.js
/**
 * Hash and sampling functions for ML-DSA (FIPS 204)
 * Uses Node.js crypto for SHAKE-128 and SHAKE-256
 */

import { createHash } from 'node:crypto';
import { Q, modQ } from './field.js';

/**
 * SHAKE-256 hash.
 * @param {Uint8Array} data - input data
 * @param {number} outLen - output length in bytes
 * @returns {Uint8Array}
 */
export function h(data, outLen) {
  const hasher = createHash('shake256', { outputLength: outLen });
  hasher.update(data);
  return new Uint8Array(hasher.digest());
}

/**
 * SHAKE-128 XOF.
 * @param {Uint8Array} data - input data
 * @param {number} outLen - output length in bytes
 * @returns {Uint8Array}
 */
function shake128(data, outLen) {
  const hasher = createHash('shake128', { outputLength: outLen });
  hasher.update(data);
  return new Uint8Array(hasher.digest());
}

/**
 * Expand the matrix A from seed rho using SHAKE-128.
 * Each element A[i][j] is a polynomial of 256 coefficients in NTT domain.
 * Rejection sampling: read 3 bytes, form 24-bit value, accept if < Q.
 * @param {Uint8Array} rho - 32-byte seed
 * @param {number} k - rows
 * @param {number} l - columns
 * @returns {Int32Array[][]} k x l matrix of polynomials
 */
export function expandA(rho, k, l) {
  const A = [];
  for (let i = 0; i < k; i++) {
    A[i] = [];
    for (let j = 0; j < l; j++) {
      A[i][j] = rejectionSampleA(rho, i, j);
    }
  }
  return A;
}

/**
 * Rejection sample a single polynomial for matrix A.
 */
function rejectionSampleA(rho, i, j) {
  // Domain separator: rho || intToLe16(j) || intToLe16(i)
  // FIPS 204 uses rho || j (low byte) || i (low byte) for the two-byte nonce
  // but actually the spec uses: rho || IntegerToBytes(s, 1) || IntegerToBytes(r, 1)
  // where s = column index (j), r = row index (i)
  const input = new Uint8Array(rho.length + 2);
  input.set(rho);
  input[rho.length] = j & 0xFF;
  input[rho.length + 1] = i & 0xFF;

  // We need exactly 256 coefficients; request enough bytes
  // Each coefficient needs ~3 bytes on average with ~Q/2^24 rejection rate
  const poly = new Int32Array(256);
  let coeffIdx = 0;
  let streamLen = 840; // Start with enough for ~256 samples

  while (coeffIdx < 256) {
    const stream = shake128(input, streamLen);
    let pos = 0;
    while (coeffIdx < 256 && pos + 2 < stream.length) {
      const b0 = stream[pos++];
      const b1 = stream[pos++];
      const b2 = stream[pos++];
      const val = b0 | (b1 << 8) | ((b2 & 0x7F) << 16);
      if (val < Q) {
        poly[coeffIdx++] = val;
      }
    }
    if (coeffIdx < 256) {
      streamLen += 840;
    }
  }
  return poly;
}

/**
 * Expand secret vectors s1, s2 from rhoPrime using SHAKE-256.
 * Coefficients are in [-eta, eta].
 * @param {Uint8Array} rhoPrime - 64-byte seed
 * @param {number} eta - bound parameter
 * @param {number} k - number of polynomials in s2
 * @param {number} l - number of polynomials in s1
 * @returns {{ s1: Int32Array[], s2: Int32Array[] }}
 */
export function expandS(rhoPrime, eta, k, l) {
  const s1 = [];
  const s2 = [];
  for (let r = 0; r < l; r++) {
    s1.push(rejectionSampleEta(rhoPrime, eta, r));
  }
  for (let r = 0; r < k; r++) {
    s2.push(rejectionSampleEta(rhoPrime, eta, l + r));
  }
  return { s1, s2 };
}

/**
 * Rejection sample a polynomial with coefficients in [-eta, eta].
 */
function rejectionSampleEta(seed, eta, nonce) {
  const input = new Uint8Array(seed.length + 2);
  input.set(seed);
  input[seed.length] = nonce & 0xFF;
  input[seed.length + 1] = (nonce >>> 8) & 0xFF;

  const poly = new Int32Array(256);
  let coeffIdx = 0;
  let streamLen = 272; // enough for most cases

  while (coeffIdx < 256) {
    const stream = h(input, streamLen);
    let pos = 0;

    if (eta === 2) {
      // Each byte gives 2 samples: low nibble and high nibble
      while (coeffIdx < 256 && pos < stream.length) {
        const b = stream[pos++];
        const t0 = b & 0x0F;
        const t1 = (b >>> 4) & 0x0F;
        if (t0 < 15) {
          poly[coeffIdx++] = 2 - (t0 % 5);
        }
        if (coeffIdx < 256 && t1 < 15) {
          poly[coeffIdx++] = 2 - (t1 % 5);
        }
      }
    } else if (eta === 4) {
      // Each byte gives 2 samples
      while (coeffIdx < 256 && pos < stream.length) {
        const b = stream[pos++];
        const t0 = b & 0x0F;
        const t1 = (b >>> 4) & 0x0F;
        if (t0 < 9) {
          poly[coeffIdx++] = 4 - t0;
        }
        if (coeffIdx < 256 && t1 < 9) {
          poly[coeffIdx++] = 4 - t1;
        }
      }
    }

    if (coeffIdx < 256) {
      streamLen += 272;
    }
  }
  return poly;
}

/**
 * Expand the mask vector y from rhoPrime using SHAKE-256.
 * Coefficients are in (-gamma1, gamma1].
 * @param {Uint8Array} rhoPP - seed (typically rho'' or rhoPrime)
 * @param {number} gamma1 - bound
 * @param {number} l - number of polynomials
 * @param {number} kappa - counter (nonce base)
 * @returns {Int32Array[]}
 */
export function expandMask(rhoPP, gamma1, l, kappa) {
  const y = [];
  for (let i = 0; i < l; i++) {
    const nonce = kappa + i;
    const input = new Uint8Array(rhoPP.length + 2);
    input.set(rhoPP);
    input[rhoPP.length] = nonce & 0xFF;
    input[rhoPP.length + 1] = (nonce >>> 8) & 0xFF;

    const poly = new Int32Array(256);

    if (gamma1 === (1 << 17)) {
      // 18 bits per coefficient => 256 * 18 / 8 = 576 bytes
      const stream = h(input, 576);
      for (let j = 0; j < 256; j++) {
        const bitOffset = j * 18;
        const byteIdx = bitOffset >>> 3;
        const bitIdx = bitOffset & 7;
        let val = (stream[byteIdx] >>> bitIdx)
          | (stream[byteIdx + 1] << (8 - bitIdx))
          | (stream[byteIdx + 2] << (16 - bitIdx));
        val &= 0x3FFFF; // 18 bits
        poly[j] = gamma1 - val;
      }
    } else if (gamma1 === (1 << 19)) {
      // 20 bits per coefficient => 256 * 20 / 8 = 640 bytes
      const stream = h(input, 640);
      for (let j = 0; j < 256; j++) {
        const bitOffset = j * 20;
        const byteIdx = bitOffset >>> 3;
        const bitIdx = bitOffset & 7;
        let val = (stream[byteIdx] >>> bitIdx)
          | (stream[byteIdx + 1] << (8 - bitIdx))
          | (stream[byteIdx + 2] << (16 - bitIdx));
        if (bitIdx > 4) {
          val |= (stream[byteIdx + 3] << (24 - bitIdx));
        }
        val &= 0xFFFFF; // 20 bits
        poly[j] = gamma1 - val;
      }
    }

    y.push(poly);
  }
  return y;
}

/**
 * Sample a polynomial c in the ball: exactly tau non-zero coefficients,
 * each +/-1, using Fisher-Yates with SHAKE-256.
 * @param {Uint8Array} cTilde - seed (lambda/4 bytes)
 * @param {number} tau - number of non-zero coefficients
 * @returns {Int32Array} polynomial with tau entries in {-1, +1}, rest 0
 */
export function sampleInBall(cTilde, tau) {
  // We need 8 bytes for signs (64 bits), plus bytes for the Fisher-Yates indices.
  // Generous buffer: tau indices with rejection sampling may need extra bytes.
  const stream = h(cTilde, 8 + 512);
  const c = new Int32Array(256);

  // First 8 bytes encode sign bits (little-endian 64-bit integer)
  let signs = BigInt(0);
  for (let i = 7; i >= 0; i--) {
    signs = (signs << BigInt(8)) | BigInt(stream[i]);
  }

  let pos = 8;
  for (let i = 256 - tau; i < 256; i++) {
    // Rejection sample: read bytes until we get j <= i
    let j = stream[pos++];
    while (j > i) {
      j = stream[pos++];
    }

    c[i] = c[j];
    c[j] = (signs & BigInt(1)) ? -1 : 1;
    signs >>= BigInt(1);
  }

  return c;
}
